### PR TITLE
[EDU-1631] Spaces Component Locking example

### PR DESCRIPTION
**Pull request by GregHolmes**
This PR contains the two code samples for the new Examples section we are creating. It contains a Typescript and a React version for Component Locking with Spaces. Which will be rendered in the code sandbox live examples in the future, along with others.

The README within each directory /examples/spaces-component-locking/javascript and /examples/spaces-component-locking/react contains information on how to get these running.

[Jira epic ticket](https://ably.atlassian.net/browse/EDU-1541)
[Jira ticket](https://ably.atlassian.net/browse/EDU-1631).

### CODE REVIEW

File: examples/spaces-component-locking/react/src/hooks/useLiveValue.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  profileData: { memberColor: string; memberName: string };
+};
+
+export const useLiveValue = (componentName: string, self: Member | null) => {
+  const [value, setValue] = useState("");
+
+  /** 💡 Use rewind to get the last message from the channel. See https://ably.com/docs/channels/options/rewind 💡 */
+  const channelName = `component-locking-${componentName}`;
+
+  console.log(channelName);
```
Comment (ttypic) [inline examples/spaces-component-locking/react/src/hooks/useLiveValue.tsx:16]: Do we need this `console.log`?

**Review by ttypic — APPROVED** [review]
LGTM

File: examples/spaces-component-locking/page.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
+# Locking components within collaborative applications
+
+Users can lock components within a collaborative application.
+
+Manage concurrent access to shared resources within your collaborative application. This feature is particularly useful for preventing conflicts in collaborative environments, ensuring that only one user can modify a specific component at a time.
```
Comment (m-hulbert) [inline examples/spaces-component-locking/page.md:5]:
```suggestion
Manage concurrent access to shared resources within your application and prevent conflicts in collaborative environments, ensuring that only one user can modify a specific component at a time.
```

File: examples/spaces-component-locking/page.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
+# Locking components within collaborative applications
+
+Users can lock components within a collaborative application.
```
Comment (m-hulbert) [inline examples/spaces-component-locking/page.md:3]:
```suggestion
Enable users to lock components for editing within a collaborative application.
```

File: examples/spaces-component-locking/page.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+Component locking are implemented using [Ably Spaces](https://ably.com/docs/products/spaces). The Spaces SDK contains a set of purpose-built APIs that abstract away the complexities involved in managing the state of components accessibility in a collaborative application. It is built on top of Ably's core platform, and so it provides the same performance and scaling guarantees.
+
+## Resources
+
+// React brief (Only visible if viewing the React example)
+
+Use the following components to add component locking into an application:
+
+* [`SpacesProvider`](https://ably.com/docs/spaces/react#spaces-provider): initializes and manages a shared space client instance, passing it down through React context to enable real-time spaces functionality across the application.
+* [`SpaceProvider`](https://ably.com/docs/spaces/react#spaces-provider): manages the state and functionality of a specific space, providing access to real-time interactions within that space via React context.
```
Comment (m-hulbert) [inline examples/spaces-component-locking/page.md:18]: real-time --> realtime

File: examples/spaces-component-locking/page.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
+# Locking components within collaborative applications
+
+Users can lock components within a collaborative application.
+
+Manage concurrent access to shared resources within your collaborative application. This feature is particularly useful for preventing conflicts in collaborative environments, ensuring that only one user can modify a specific component at a time.
+
+Component locking enhances various collaborative applications, such as, in document editing tools, it can prevent simultaneous edits to the same section. In project management software, it can ensure that task details are updated by one team member at a time, maintaining data integrity and preventing conflicting changes.
+
+Component locking are implemented using [Ably Spaces](https://ably.com/docs/products/spaces). The Spaces SDK contains a set of purpose-built APIs that abstract away the complexities involved in managing the state of components accessibility in a collaborative application. It is built on top of Ably's core platform, and so it provides the same performance and scaling guarantees.
```
Comment (m-hulbert) [inline examples/spaces-component-locking/page.md:9]:
```suggestion
Component locking is implemented using [Ably Spaces](https://ably.com/docs/products/spaces). The Spaces SDK contains a set of purpose-built APIs that abstract away the complexities involved in managing the state of components accessibility in a collaborative application. It is built on top of Ably's core platform, and so it provides the same performance and scaling guarantees.
```

File: examples/spaces-component-locking/page.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
+# Locking components within collaborative applications
+
+Users can lock components within a collaborative application.
+
+Manage concurrent access to shared resources within your collaborative application. This feature is particularly useful for preventing conflicts in collaborative environments, ensuring that only one user can modify a specific component at a time.
+
+Component locking enhances various collaborative applications, such as, in document editing tools, it can prevent simultaneous edits to the same section. In project management software, it can ensure that task details are updated by one team member at a time, maintaining data integrity and preventing conflicting changes.
```
Comment (m-hulbert) [inline examples/spaces-component-locking/page.md:7]:
```suggestion
Component locking enhances applications where multiple users can edit the same fields, such as spreadsheets, slide shows and forms. It is a key component in maintaining data integrity and preventing conflicting changes.
```

**Review by m-hulbert — CHANGES_REQUESTED** [review]
If you remove focus (e.g. click on your other window) then your component is cleared. This is different behaviour to the current example, and I think we need to keep it in order to display the functionality correctly. (JS)

The React example has the issue of reading its CSS from `public` still.

**Review by m-hulbert — COMMENTED** [review]
This all works perfectly now. 

I think we need to add the `ChannelProvider` to the React entry and `publish()` to the JS one.

I also think it might be worth calling out the providers / methods that we're using from Pub/Sub separately (just a sentence rather than another section header). Thoughts?

**Review by m-hulbert — APPROVED** [review]

**Pull request merged**